File: cpp/src/chips/pressure/BMP180.cpp
#include "BMP180.h"
#include <stdlib.h>

BMP180Minimal::BMP180Minimal(Transport& transport)
    : _transport(transport) {
    _read_calibration();
}

void BMP180Minimal::_read_calibration() {
    uint8_t buf[22];
    uint8_t reg = REG_CAL_START;
    _transport.write_read(&reg, 1, buf, 22);

    _ac1 = (int16_t)((buf[0] << 8) | buf[1]);
    _ac2 = (int16_t)((buf[2] << 8) | buf[3]);
    _ac3 = (int16_t)((buf[4] << 8) | buf[5]);
    _ac4 = (uint16_t)((buf[6] << 8) | buf[7]);
    _ac5 = (uint16_t)((buf[8] << 8) | buf[9]);
    _ac6 = (uint16_t)((buf[10] << 8) | buf[11]);
    _b1  = (int16_t)((buf[12] << 8) | buf[13]);
    _b2  = (int16_t)((buf[14] << 8) | buf[15]);
    _mb  = (int16_t)((buf[16] << 8) | buf[17]);
    _mc  = (int16_t)((buf[18] << 8) | buf[19]);
    _md  = (int16_t)((buf[20] << 8) | buf[21]);

    if (_ac1 == 0 || _ac1 == 0xFFFF ||
        _ac2 == 0 || _ac2 == 0xFFFF ||
        _ac3 == 0 || _ac3 == 0xFFFF ||
        _ac4 == 0 || _ac4 == 0xFFFF ||
        _ac5 == 0 || _ac5 == 0xFFFF ||
        _ac6 == 0 || _ac6 == 0xFFFF ||
        _b1  == 0 || _b1  == 0xFFFF ||
        _b2  == 0 || _b2  == 0xFFFF ||
        _mb  == 0 || _mb  == 0xFFFF ||
        _mc  == 0 || _mc  == 0xFFFF ||
        _md  == 0 || _md  == 0xFFFF) {
        abort();
    }
}

void BMP180Minimal::_write_reg(uint8_t reg, uint8_t value) {
    uint8_t buf[2] = { reg, value };
    _transport.write(buf, 2);
}

uint16_t BMP180Minimal::_read_raw_temp() {
    _write_reg(REG_CTRL_MEAS, CMD_TEMP);
    delay(CONV_TIME_TEMP * 1000);
    uint8_t buf[2];
    uint8_t reg = REG_OUT_MSB;
    _transport.write_read(&reg, 1, buf, 2);
    return ((uint16_t)buf[0] << 8) | buf[1];
}

uint32_t BMP180Minimal::_read_raw_pressure() {
    uint8_t cmd = (_oss == 0) ? CMD_PRESSURE_OSS0 :
                  (_oss == 1) ? CMD_PRESSURE_OSS1 :
                  (_oss == 2) ? CMD_PRESSURE_OSS2 : CMD_PRESSURE_OSS3;
    float conv_time = (_oss == 0) ? CONV_TIME_OSS0 :
                      (_oss == 1) ? CONV_TIME_OSS1 :
                      (_oss == 2) ? CONV_TIME_OSS2 : CONV_TIME_OSS3;
    _write_reg(REG_CTRL_MEAS, cmd);
    delay(conv_time * 1000);
    uint8_t buf[3];
    uint8_t reg = REG_OUT_MSB;
    _transport.write_read(&reg, 1, buf, 3);
    uint32_t up = (((uint32_t)buf[0] << 16) | ((uint32_t)buf[1] << 8) | buf[2]) >> (8 - _oss);
    return up;
}

int32_t BMP180Minimal::_compensate_temp(int32_t ut) {
    int32_t x1 = ((ut - (int32_t)_ac6) * (int32_t)_ac5) >> 15;
    int32_t x2 = ((int32_t)_mc << 11) / (x1 + (int32_t)_md);
    int32_t b5 = x1 + x2;
    _b5 = b5;
    return b5;
}

int32_t BMP180Minimal::_compensate_pressure(int32_t up) {
    int32_t oss = (int32_t)_oss;
    int32_t b6 = _b5 - 4000;
    int32_t x1 = ((int32_t)_b2 * ((b6 * b6) >> 12)) >> 11;
    int32_t x2 = ((int32_t)_ac2 * b6) >> 11;
    int32_t x3 = x1 + x2;
    int32_t b3 = (((((int32_t)_ac1 * 4) + x3) << oss) + 2) >> 2;
    x1 = ((int32_t)_ac3 * b6) >> 13;
    x2 = ((int32_t)_b1 * ((b6 * b6) >> 12)) >> 16;
    x3 = ((x1 + x2) + 2) >> 2;
    uint32_t b4 = ((uint32_t)_ac4 * ((uint32_t)(x3 + 32768))) >> 15;
    uint32_t b7 = ((uint32_t)(up - b3)) * ((uint32_t)(50000 >> oss));

    int32_t p;
    if (b7 < 0x80000000) {
        p = (int32_t)(((uint32_t)(b7 * 2)) / b4);
    } else {
        p = (int32_t)(((b7 / (int32_t)b4) * 2));
    }

    x1 = (p >> 8) * (p >> 8);
    x1 = (x1 * 3038) >> 16;
    x2 = (-7357 * p) >> 16;
    p = p + ((x1 + x2 + 3791) >> 4);

    return p;
}

float BMP180Minimal::temperature() {
    uint16_t ut = _read_raw_temp();
    int32_t b5 = _compensate_temp(ut);
    return ((float)(b5 + 8)) / 160.0f;
}

float BMP180Minimal::pressure() {
    uint16_t ut = _read_raw_temp();
    int32_t b5 = _compensate_temp(ut);
    (void)b5;
    uint32_t up = _read_raw_pressure();
    int32_t p_pa = _compensate_pressure((int32_t)up);
    return ((float)p_pa) / 100.0f;
}

// BMP180Full

BMP180Full::BMP180Full(Transport& transport, uint8_t oss)
    : BMP180Minimal(transport) {
    _oss = oss & 0x03;
}

uint8_t BMP180Full::oversampling() {
    return _oss;
}

void BMP180Full::set_oversampling(uint8_t oss) {
    _oss = oss & 0x03;
}

float BMP180Full::altitude(float sea_level_hpa) {
    float p = pressure();
    return 44330.0f * (1.0f - powf(p / sea_level_hpa, 0.1903f));
}

float BMP180Full::sea_level_pressure(float altitude_m) {
    float p = pressure();
    return p / powf(1.0f - ((float)altitude_m / 44330.0f), 5.255f);
}

uint8_t BMP180Full::chip_id() {
    uint8_t reg = REG_ID;
    uint8_t buf[1];
    _transport.write_read(&reg, 1, buf, 1);
    return buf[0];
}

void BMP180Full::reset() {
    _write_reg(REG_SOFT_RESET, SOFT_RESET_CMD);
    delay(10);
    _read_calibration();
}

File: nodejs/packages/periph/src/chips/pressure/bmp180.js
'use strict';

const _REG_ID         = 0xD0;
const _REG_CAL_START = 0xAA;
const _REG_CTRL_MEAS = 0xF4;
const _REG_OUT_MSB   = 0xF6;
const _REG_OUT_LSB   = 0xF7;
const _REG_OUT_XLSB  = 0xF8;
const _REG_SOFT_RESET = 0xE0;

const _CMD_TEMP = 0x2E;
const _CMD_PRESSURE = [0x34, 0x74, 0xB4, 0xF4];

const _CHIP_ID       = 0x55;
const _SOFT_RESET_CMD = 0xB6;

const _CONV_TIME = [0.0045, 0.0075, 0.0135, 0.0255];
const _CONV_TIME_TEMP = 0.0045;

function _delay(ms) {
    const start = Date.now();
    while (Date.now() - start < ms) { /* spin */ }
}

/**
 * BMP180 piezo-resistive pressure + temperature sensor — minimal interface.
 *
 * Provides calibrated temperature (°C) and pressure (hPa) with no configuration
 * beyond the transport. Fixed I²C address 0x77.
 *
 * Default OSS = 0 (Ultra Low Power, 4.5 ms conversion).
 *
 * @param {object} transport - Configured I²C transport.
 */
class BMP180Minimal {
    constructor(transport) {
        this._transport = transport;
        this._oss = 0;
        this._readCalibration();
    }

    _readCalibration() {
        const data = this._transport.writeRead(Buffer.from([_REG_CAL_START]), 22);
        this._ac1 = data.readInt16BE(0);
        this._ac2 = data.readInt16BE(2);
        this._ac3 = data.readInt16BE(4);
        this._ac4 = data.readUInt16BE(6);
        this._ac5 = data.readUInt16BE(8);
        this._ac6 = data.readUInt16BE(10);
        this._b1  = data.readInt16BE(12);
        this._b2  = data.readInt16BE(14);
        this._mb  = data.readInt16BE(16);
        this._mc  = data.readInt16BE(18);
        this._md  = data.readInt16BE(20);

        const coefficients = [this._ac1, this._ac2, this._ac3, this._ac4,
            this._ac5, this._ac6, this._b1, this._b2, this._mb, this._mc, this._md];
        if (coefficients.some(c => c === 0 || c === 0xFFFF)) {
            throw new Error('BMP180 calibration data invalid');
        }
    }

    _writeReg(reg, value) {
        this._transport.write(Buffer.from([reg, value]));
    }

    _readRawTemp() {
        this._writeReg(_REG_CTRL_MEAS, _CMD_TEMP);
        _delay(_CONV_TIME_TEMP * 1000);
        const data = this._transport.writeRead(Buffer.from([_REG_OUT_MSB]), 2);
        return data.readUInt16BE(0);
    }

    _readRawPressure() {
        const cmd = _CMD_PRESSURE[this._oss];
        this._writeReg(_REG_CTRL_MEAS, cmd);
        _delay(_CONV_TIME[this._oss] * 1000);
        const data = this._transport.writeRead(Buffer.from([_REG_OUT_MSB]), 3);
        let up = ((data.readUInt16BE(0) << 8) | data[2]) >> (8 - this._oss);
        return up;
    }

    _compensateTemp(ut) {
        const ac1 = this._ac1, ac2 = this._ac2, ac3 = this._ac3;
        const ac4 = this._ac4, ac5 = this._ac5, ac6 = this._ac6;
        const b1 = this._b1, b2 = this._b2;
        const mc = this._mc, md = this._md;

        let x1 = Math.floor(((ut - ac6) * ac5) / 32768);
        let x2 = Math.floor((mc << 11) / (x1 + md));
        let b5 = x1 + x2;
        this._b5 = b5;
        return b5;
    }

    _compensatePressure(up) {
        const oss = this._oss;
        const ac1 = this._ac1, ac2 = this._ac2, ac3 = this._ac3;
        const ac4 = this._ac4;
        const b1 = this._b1, b2 = this._b2;
        const b5 = this._b5;

        let b6 = b5 - 4000;
        let x1 = Math.floor((b2 * Math.floor((b6 * b6) / 4096)) / 2048);
        let x2 = Math.floor((ac2 * b6) / 2048);
        let x3 = x1 + x2;
        let b3 = Math.floor((((ac1 * 4) + x3) << oss) / 4);
        x1 = Math.floor((ac3 * b6) / 8192);
        x2 = Math.floor((b1 * Math.floor((b6 * b6) / 4096)) / 65536);
        x3 = Math.floor((x1 + x2 + 2) / 4);
        let b4 = Math.floor((ac4 * (x3 + 32768)) / 32768);
        let b7 = Math.floor((up - b3) * (50000 >>> oss));

        let p;
        if (b7 < 0x80000000) {
            p = Math.floor((b7 * 2) / b4);
        } else {
            p = Math.floor((b7 / b4) * 2);
        }

        x1 = Math.floor((p >> 8) * (p >> 8));
        x1 = Math.floor((x1 * 3038) / 65536);
        x2 = Math.floor((-7357 * p) / 65536);
        p = p + Math.floor((x1 + x2 + 3791) / 16);

        return p;
    }

    /**
     * Read calibrated temperature.
     * @returns {number} Temperature in degrees Celsius.
     */
    temperature() {
        const ut = this._readRawTemp();
        const b5 = this._compensateTemp(ut);
        return ((b5 + 8) / 160.0);
    }

    /**
     * Read calibrated pressure.
     *
     * Reads temperature first to refresh B5, then reads pressure.
     * Self-contained — may be called without a prior temperature() call.
     *
     * @returns {number} Pressure in hPa.
     */
    pressure() {
        const ut = this._readRawTemp();
        const b5 = this._compensateTemp(ut);
        (void)b5;
        const up = this._readRawPressure();
        const p_pa = this._compensatePressure(up);
        return p_pa / 100.0;
    }
}

/**
 * BMP180 full interface — extends BMP180Minimal with OSS control and altitude helpers.
 *
 * OSS constants:
 * - BMP180Full.OSS_ULP — Ultra Low Power (oss=0, 4.5 ms)
 * - BMP180Full.OSS_STANDARD — Standard (oss=1, 7.5 ms)
 * - BMP180Full.OSS_HIGH_RES — High Resolution (oss=2, 13.5 ms)
 * - BMP180Full.OSS_ULTRA_HIGH_RES — Ultra High Resolution (oss=3, 25.5 ms)
 *
 * @param {object} transport - Configured I²C transport.
 * @param {number} [oss=0] - Oversampling mode 0–3.
 */
class BMP180Full extends BMP180Minimal {
    static OSS_ULP            = 0;
    static OSS_STANDARD        = 1;
    static OSS_HIGH_RES        = 2;
    static OSS_ULTRA_HIGH_RES = 3;

    constructor(transport, oss = 0) {
        super(transport);
        this._oss = oss & 0x03;
    }

    /**
     * Read the current oversampling mode.
     * @returns {number} OSS value 0–3.
     */
    oversampling() {
        return this._oss;
    }

    /**
     * Change the oversampling mode for subsequent pressure() calls.
     * @param {number} oss - New OSS value 0–3.
     */
    setOversampling(oss) {
        this._oss = oss & 0x03;
    }

    /**
     * Compute altitude above sea level from the current pressure.
     * @param {number} [seaLevelHpa=1013.25] - Reference sea-level pressure in hPa.
     * @returns {number} Altitude in metres.
     */
    altitude(seaLevelHpa = 1013.25) {
        const p = this.pressure();
        return 44330 * (1 - Math.pow(p / seaLevelHpa, 1 / 5.255));
    }

    /**
     * Compute sea-level pressure for a known altitude.
     * @param {number} altitudeM - Altitude in metres.
     * @returns {number} Sea-level pressure in hPa.
     */
    seaLevelPressure(altitudeM) {
        const p = this.pressure();
        return p / Math.pow(1 - altitudeM / 44330, 5.255);
    }

    /**
     * Read the chip ID register.
     * @returns {number} Chip ID; expect 0x55.
     */
    chipId() {
        const data = this._transport.writeRead(Buffer.from([_REG_ID]), 1);
        return data[0];
    }

    /**
     * Perform a soft reset and re-read calibration coefficients.
     */
    reset() {
        this._writeReg(_REG_SOFT_RESET, _SOFT_RESET_CMD);
        _delay(10);
        this._readCalibration();
    }
}

module.exports = { BMP180Minimal, BMP180Full };
